wall_e/extensions/sfu.py

```diff
@@ -502,6 +506,104 @@ async def outline(self, ctx, *course):
+                res_json = await res.json()
+
+                # parse data for displaying, sorting, and filtering
+                for course in res_json:
+                    course["text"] = f"{department}{course['text']}"
+
+                    # we assume all courses have 3 digits
+                    if len(course["value"]) == 4:
+                        course["value"] = course["value"][:3]
+
+                    course["value"] = int(course["value"])
+                    if level == 0 or (course["value"]//100 == level//100):
+                        courses.append(course)
+            else:
+                self.logger.debug(f"[SFU courses()] get request for {department} resulted in {res.status}")
+
+        if len(courses) == 0:
+            self.logger.debug("[SFU courses()] resulted in no content")
+            e_obj = await embed(
+                self.logger, interaction=interaction, title="SFU Courses Error",
+                description=(
+                    f"Couldn't find anything for `department: {', '.join(departments)}"
+                    f", level: {level if level != 0 else 'all'}, term: {term}, year: {year}`"
+                    f"\n Maybe no courses are being offered at that time."
+                ),
+                colour=WallEColour.ERROR,
+            )
+            if e_obj:
+                msg = await interaction.followup.send(embed=e_obj)
+                await asyncio.sleep(10)
+                await msg.delete()
+            return
+
+        if len(departments) > 1:
+            courses = sorted(courses, key=lambda k: k["value"])
+
+        self.logger.debug("[SFU courses()] parsing data from get request")
+        content_to_embed = []
+        number_of_courses_per_page = 20
+        number_of_courses = 0
+        total_course = 0
+        content = ""
+
+        for i, course in enumerate(courses):
+            course_title = course["title"]
+            course_number = course["text"]
+            content += f"\n{course_number} - {course_title}"
+
+            total_course += 1
+            number_of_courses += 1
+            if total_course > 0 and (number_of_courses % number_of_courses_per_page == 0 or i == len(courses) - 1):
+                number_of_courses = 0
+                content_to_embed.append(
+                    [["Code - Title", content]]
+                )
+                content = ""
+
+        title = (f"{', '.join(departments)} {f'{(level//100)*100} level -' if level != 0 else '-'}"
+                 f" {f'{term.title()}' if term != 'registration' and term != 'current' else f'{term} term'}"
+                 f" {f'{year.title()}' if year != 'registration' and year != 'current' else f'{year} year'}"
+                 f"\n(Total Courses: {total_course})\n")
+
+        await paginate_embed(
+            self.logger, bot, content_to_embed=content_to_embed,
+            title=title,
+            interaction=interaction
+        )
```

Comment:
The formatting chunk of the command looks like a good candidate to refactor into it's own function! 

Something like `async def _create_courses_paginate_embed(courses: list): ...`, with an underscore alluding that it's not called by users (private)

Reply:
just curious, what is the criteria for a chunk of code to be refactored
for example with the code above, the pulling of data from sfu, could that also be a candidate to refactor?

Reply:
also that function would need to rely on `departments`, `level`, `term`, and `year` as well, would it be ok to add those?

Reply:
> what is the criteria for a chunk of code to be refactored

The criteria is usually:
1. when there's a large block of code
2. a section of the code is lowly coupled with the rest of the function (essentially means it depends on few variables above it)
3. it has a specific purpose.

Deciding when to break code into chunks has a lot of exceptions, but the above rules are a somewhat good approximation. The goal is for code to be easy to read by new developers!

> that function would need to rely on `departments`, `level`, `term`, and `year` as well, would it be ok to add those?

Ah, good point! I didn't notice that. Yep, it would be okay.

However given this information we have another option; perhaps just lines 575 to line 594 should be turned into a function which depends on `courses` and returns `content_to_embed`? This has a lower coupling than if we made the entire formatting section into a function. Perhaps you can see where the line starts to fuzzy and debatable?

Feel free to choose whichever you think makes the code best in your opinion! As long as you've thought it through I'll be happy.

Reply:
> that function would need to rely on `departments`, `level`, `term`, and `year` as well, would it be ok to add those?

Ah, good point! I didn't notice that. Yep, it would be okay.

However given this information we have another option; perhaps just lines 575 to line 594 should be turned into a function which depends on `courses` and returns `content_to_embed`? This has a lower coupling than if we made the entire formatting section into a function. Perhaps you can see where the line starts to fuzzy and debatable?

Feel free to choose whichever you think makes the code best in your opinion! As long as you've thought it through I'll be happy.

Reply:
https://github.com/CSSS/wall_e/pull/666/commits/5f54299272d64864628b480c63f3459a4236650e